### PR TITLE
Add note about tests with randomness to the contribution guide.

### PR DESCRIPTION
**Pull request by Ericgig**
**Description**
Add a section in contributor section about how to handle randomness in tests.

### CODE REVIEW

File: doc/development/contribution_guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+Testing guidelines: non-deterministic tests
+===========================================
+
+In QuTiP's test there are two common sources of randomness in tests:
+
+- Random inputs:
+
+  The functions tested is itself deterministic, but the inputs are generated randomly.
+  It is generally good practice to include some tests with random inputs as these
+  can allows to find unexpected edge cases. **Do no fix the seed**, instead
```
Comment (hodgestar) [inline doc/development/contribution_guidelines.rst]:
```suggestion
  can allows to find unexpected edge cases. **Do not fix the seed**, instead
```

File: doc/development/contribution_guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+  tolerance issue.
+
+
+- Handling Edge Cases:
+
+  - Logic Errors in the Function:
+
+    If the failure comes from an edge cases that is not, but should be supported,
+    please fix it if it is within the scope of your contribution. Otherwise add an
+    ``XFAIL`` test as and open an GitHub issue for a future fix.
```
Comment (hodgestar) [inline doc/development/contribution_guidelines.rst]:
```suggestion
    ``XFAIL`` test and open an GitHub issue for a future fix.
```

File: doc/development/contribution_guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+    Example:
+
+      Monte Carlo evolution could detect a "jump" after reaching the ground state
+      due to the finite precision of the ODE solver instead of the rather than
+      the underlying physics.
```
Comment (hodgestar) [inline doc/development/contribution_guidelines.rst:47]:
```suggestion
      Monte Carlo evolution could detect a "jump" after reaching the ground state
      due to the finite precision of the ODE solver instead of the
      the underlying physics.
```

File: doc/development/contribution_guidelines.rst [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+      The solution would be to scale the random operator to ensure it stays
+      within a valid physical range.
+
+
+Numerical Tolerance Issues:
+
+  Sometime, either through accumulation of numerical error or finite precision
+  of iterative method, a test may fail occasionally if the tolerance is too tight.
+  In this situation estimate the distribution of the result over multiple runs
+  and set the tolerance to 4-sigma level (failing less than 1 in 10000).
```
Comment (hodgestar) [inline doc/development/contribution_guidelines.rst]:
```suggestion
  Sometimes, either through accumulation of numerical error or the finite precision
  of an iterative method, a test may fail occasionally if the tolerance is too tight.
  In this situation, estimate the distribution of the result over multiple runs
  and set the tolerance to 4-sigma level (failing less than 1 in 10000).
```

File: doc/development/contribution_guidelines.rst [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+.. note::
+
+  A Note on Floating Point Non-determinism:
+
+  Even if nothing use random number, some computation can return slightly different
+  results with across calls. This is usually caused by floating-point arithmetic
+  ``(A + B) + C != A + (B + C)`` and asynchronous computation associated with
```
Comment (hodgestar) [inline doc/development/contribution_guidelines.rst:74]:
```suggestion
  A Note on Floating Point Non-determinism:

  Even if nothing uses random numbers, some computations may return slightly different
  results across calls. This is usually caused by floating-point arithmetic
```

**Review by hodgestar — CHANGES_REQUESTED** [review]
Looks good. I left some small suggestions.

File: doc/development/contribution_guidelines.rst (hunk 5 above)
Comment (Ericgig) [reply doc/development/contribution_guidelines.rst:74]:
```suggestion
  and asynchronous computation associated with
```
?

Comment (Ericgig) [reply doc/development/contribution_guidelines.rst:74]: @hodgestar Your suggestion did not show properly, is my understanding right?

Comment (hodgestar) [reply doc/development/contribution_guidelines.rst:74]: Let me just close this suggestion and re-read the text. I can't understand what the diff is trying to do either.

File: doc/development/contribution_guidelines.rst [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  parallel maps.  These should generally be treated as tolerance issues.
```
Comment (hodgestar) [inline doc/development/contribution_guidelines.rst]:
```suggestion
  Even if nothing uses random number, some computations may return slightly different
  results across calls. This is usually caused by floating-point arithmetic
  ``(A + B) + C != A + (B + C)`` and asynchronous computation associated with
  parallel maps.  These case should be treated as tolerance issues.
```

**Review by hodgestar — APPROVED** [review]

**Comment by hodgestar** [comment]
@Ericgig I redid the suggestion and also approved in case you want to merge straight after applying the suggestion.
